posts/2024-10-21-data.table.threads-anirban_chetia/index.qmd

```diff
@@ -0,0 +1,54 @@
+
+Printing this object would enlist the fastest median runtime (in milliseconds) along with the thread count that achieved it for each function.
+
+Plotting the object would generate a plot that shows the ideal and measured speedup trends for each benchmarked `data.table` function.
+
+If the user wants to factor in a specified speedup efficiency, they can use the function `addRecommendedEfficiency` to add a speedup line (with a slope configured by input argument `efficiencyFactor`; default value is 0.5, or 50% efficiency) along with a point representing the recommended thread count which stems from the highest intersection between this line (of specified thread-use efficiency) and measured speedup data for each function.
+
+Here is an example:
+```{r}
+library(data.table.threads)
+
+(benchmarks <- data.table.threads::findOptimalThreadCount(1e7, 10, verbose = FALSE))
+plot(benchmarks)
+benchmarks_r <- addRecommendedEfficiency(benchmarks, recommendedEfficiency = 0.6)
+plot(benchmarks_r)
+```
+
+In order to set the thread count based on observed results for a user-specified function and speedup efficiency value, the `setThreadCount(benchmarkData, functionName, efficiencyFactor)` function can then be used:
+```{r}
+setThreadCount(benchmarks, functionName = "forder", efficiencyFactor = 1)
+```
+
+When using `findOptimalThreadCount()`, users can also replace the predefined benchmarks with their own expressions by providing a list of custom functions as the `benchmarksList` argument, enabling evaluation tailored to their specific use cases. On top of that, they can also specify their own `data.table` via the `customDT` argument to have the functions they define operate on it instead of the default matrix-based `data.table` that makes use of `rowCount` and `colCount`.
+
+For instance, here is an example:
+```{r}
+# Derived from https://github.com/Rdatatable/data.table/issues/4294
+NN = 1e5
+DT = data.table(grp1 = as.character(rep(1:(NN / 4), each = 4)),
+                grp2 = sample(5000L, NN, TRUE),
+                V = rpois(NN, 10))
+findOptimalThreadCount(benchmarksList = list(test4294 = function(dt) dt[, log(sum(V)), by = grp1]), customDT = DT)
+# Optimal efficiency is observed for a single thread in this case since the code includes creation of many small groups where the computation for each of them is relatively lightweight to see improvement in performance outweighing the overhead from multi-threading.
```

Comment:
please move comment outside of code block, and add a conclusion

Reply:
Done on both suggestions a while ago - Thanks for reviewing this! (and also my other PR/post)

Reply:
Thanks for these helpful suggestions early on! (for both this and my other PR/post)

I've made those changes and more + popped out of draft mode for both PRs (in case you would like to have another look/review)